File: day04/main.go
package main

import (
	_ "embed"
	"log/slog"
	"strings"
)

//go:embed inp
var inp string

func main() {
	slog.Info("Part 1:", "Mul:", new(solver).prep(inp).part1())
	slog.Info("Part 2:", "Mul:", new(solver).prep(inp).part2())
}

type solver struct {
	mtx [][]rune
}

func (s *solver) prep(inp string) *solver {
	rows := strings.Split(inp, "\n")
	s.mtx = make([][]rune, 0, len(rows))

	for r := range rows {
		rowStr := strings.Split(rows[r], "")
		newRow := make([]rune, 0, len(rowStr))

		for i := range rowStr {
			newRow = append(newRow, rune(rowStr[i][0]))
		}

		s.mtx = append(s.mtx, newRow)
	}

	return s
}

func (s *solver) part1() int {
	counter := 0
	for i := range len(s.mtx) {
		for j := range len(s.mtx[i]) {
			if s.mtx[i][j] == 'X' {
				counter += s.countXMAS(i, j)
			}
		}
	}

	return counter
}

func (s *solver) part2() int {
	counter := 0

	for i := range len(s.mtx) {
		for j := range len(s.mtx[i]) {
			if s.mtx[i][j] == 'A' {
				if s.countMAS(i, j, SMMS) ||
					s.countMAS(i, j, MSSM) ||
					s.countMAS(i, j, MMSS) ||
					s.countMAS(i, j, SSMM) {
					counter++
				}
			}
		}
	}

	return counter
}

type masPosSolver struct {
	target map[[2]int]rune
}

func (s *solver) countMAS(r, c int, mp masPosSolver) bool {
	for _, point := range masBounds {
		cr, cc := r+point[0], c+point[1]
		if !s.isInBound(cr, cc) {
			return false
		}

		if s.mtx[cr][cc] != mp.target[point] {
			return false
		}
	}

	return true
}

func (s *solver) countXMAS(r, c int) int {
	found := 0

	for _, dir := range xmasPoints {
		for i := 0; i < len(xmas); i++ {
			cr, cc := r+i*dir[0], c+i*dir[1]

			inc := s.countXMASInDir(i, cr, cc)
			if inc == -1 {
				break
			}

			found += inc
		}
	}

	return found
}

func (s *solver) isInBound(r, c int) bool {
	return r >= 0 && c >= 0 && r < len(s.mtx) && c < len(s.mtx[0])
}

func (s *solver) countXMASInDir(i, cr, cc int) int {
	if !s.isInBound(cr, cc) || s.mtx[cr][cc] != xmas[i] {
		return -1
	}

	if i == len(xmas)-1 && s.mtx[cr][cc] == xmas[len(xmas)-1] {
		return 1
	}

	return 0
}

// S.M
// .A.
// S.M
var SMMS = masPosSolver{
	target: map[[2]int]rune{
		{-1, -1}: 'S',
		{-1, 1}:  'M',
		{1, 1}:   'M',
		{1, -1}:  'S',
	},
}

// M.S
// .A.
// M.S
var MSSM = masPosSolver{
	target: map[[2]int]rune{
		{-1, -1}: 'M',
		{-1, 1}:  'S',
		{1, 1}:   'S',
		{1, -1}:  'M',
	},
}

// M.M
// .A.
// S.S
var MMSS = masPosSolver{
	target: map[[2]int]rune{
		{-1, -1}: 'M',
		{-1, 1}:  'M',
		{1, 1}:   'S',
		{1, -1}:  'S',
	},
}

// S.S
// .A.
// M.M
var SSMM = masPosSolver{
	target: map[[2]int]rune{
		{-1, -1}: 'S',
		{-1, 1}:  'S',
		{1, 1}:   'M',
		{1, -1}:  'M',
	},
}

var (
	xmas       = []rune{'X', 'M', 'A', 'S'}
	xmasPoints = [][2]int{
		{1, 0},   // down
		{0, 1},   // right
		{-1, 0},  // up
		{0, -1},  // left
		{1, 1},   // down-right
		{-1, 1},  // up-right
		{-1, -1}, // up-left
		{1, -1},  // down-left
	}
	masBounds = [][2]int{
		{-1, -1}, // up-left
		{-1, 1},  // up-right
		{1, 1},   // down-right
		{1, -1},  // down-left
	}
)

File: day04/src/main.rs
mod main_test;

use std::fs;
use std::ops::Neg;

fn main() {
    let content = fs::read_to_string("./day04/inp");
    match content {
        Ok(content) => {
            let s: Solver = Solver::new().prep(&content);
            println!("{:?}", s.part1());
            println!("{:?}", s.part2());
        }
        Err(e) => {
            println!("{:?}", e);
        }
    }
}

struct Solver {
    mtx: Vec<Vec<char>>,
}

impl Solver {
    fn new() -> Solver {
        Solver { mtx: vec![] }
    }

    fn prep(mut self, inp: &str) -> Self {
        self.mtx = inp
            .lines()
            .map(|line| line.chars().collect::<Vec<char>>())
            .collect();
        self
    }

    fn part1(&self) -> i64 {
        self.mtx
            .iter()
            .enumerate()
            .flat_map(|(i, row)| {
                row.iter().enumerate().filter_map(move |(j, &cell)| {
                    if cell == 'X' {
                        Some(self.count_xmas(i, j))
                    } else {
                        None
                    }
                })
            })
            .sum()
    }

    fn part2(&self) -> i64 {
        0
    }

    fn count_xmas(&self, row: usize, col: usize) -> i64 {
        let mut found: i64 = 0;

        for dir in XMAS_POINTS {
            for i in 0..XMAS.len() {
                let letter = &XMAS[i];
                let step_row: usize;
                let step_col: usize;

                if let Some([new_row, new_col]) = self.new_row_col(dir, i, row, col) {
                    step_row = new_row;
                    step_col = new_col;
                } else {
                    break;
                }

                if !self.is_in_bound(step_row, step_col)
                    || self.mtx[step_row][step_col].cmp(letter).is_ne()
                {
                    break;
                }

                if *letter == 'S'
                    && self.mtx[step_row][step_col]
                        .cmp(XMAS.last().unwrap())
                        .is_eq()
                {
                    found += 1;
                }
            }
        }

        found
    }

    fn new_row_col(&self, dir: [i64; 2], i: usize, row: usize, col: usize) -> Option<[usize; 2]> {
        let row_offset = dir[0];
        let col_offset = dir[1];

        let test_cur_row: Option<usize> = if row_offset < 0 {
            row.checked_sub(i * row_offset.neg() as usize)
        } else {
            row.checked_add(i * row_offset as usize)
        };

        let test_cur_col: Option<usize> = if col_offset < 0 {
            col.checked_sub(i * col_offset.neg() as usize)
        } else {
            col.checked_add(i * col_offset as usize)
        };

        if test_cur_row.is_none() || test_cur_col.is_none() {
            return None;
        }

        Some(<[usize; 2]>::try_from(vec![test_cur_row.unwrap(), test_cur_col.unwrap()]).unwrap())
    }

    fn is_in_bound(&self, row: usize, col: usize) -> bool {
        row < self.mtx.len() && col < self.mtx[row].len()
    }
}

const XMAS: [char; 4] = ['X', 'M', 'A', 'S'];
const XMAS_POINTS: [[i64; 2]; 8] = [
    [1, 0],   // down
    [0, 1],   // right
    [-1, 0],  // up
    [0, -1],  // left
    [1, 1],   // down-right
    [-1, 1],  // up-right
    [-1, -1], // up-left
    [1, -1],  // down-left[
];

#[allow(dead_code)]
const MAS_BOUNDS: [[i64; 2]; 4] = [[-1, -1], [-1, 1], [1, 1], [1, -1]];
